src/OpenLoco/Paint/PaintIndustry.cpp

```diff
@@ -0,0 +1,74 @@
+#include "PaintIndustry.h"
+#include "../Industry.h"
+#include "../Map/Tile.h"
+#include "../Objects/IndustryObject.h"
+#include "../ScenarioManager.h"
+#include "../Ui.h"
+#include "Paint.h"
+
+namespace OpenLoco::Paint
+{
+    // 0x00453C52
+    void paintIndustry(PaintSession& session, const Map::IndustryElement& elIndustry)
+    {
+        session.setItemType(Ui::ViewportInteraction::InteractionItem::industry);
+        auto* industry = elIndustry.industry();
+        auto* indObj = industry->getObject();
+
+        ImageId baseColour(0, elIndustry.var_6_F800());
+        if (elIndustry.isGhost())
+        {
+            baseColour = Gfx::applyGhostToImage(0);
+        }
+        // Combine this with any imageId
+        const uint8_t rotation = (session.getRotation() + elIndustry.unkDirection()) & 0x3;
+        // 0xE0C3A0
+        auto ticks = ScenarioManager::getScenarioTicks();
+        uint8_t bl = 0xF0;
+        uint8_t bh = 0;
+        if (!elIndustry.hasHighTypeFlag())
+        {
+            ticks = 0;
+            bl = elIndustry.var_6_003F();
+            bh = elIndustry.var_5_E0();
+        }
+
+        // 0x00525D5C
+        uint8_t unk525D5C = elIndustry.var_5_03();
+
+        // 0x00525D4E
+        const int16_t bbLengthZ = std::min(elIndustry.clearHeight() - elIndustry.baseHeight(), 128) - 2;
+
+        // 0x00E0C3B0 (pointer to something like x,y bytes)
+        uint32_t unkE0C3B0 = 0;
+        if ((elIndustry.var_6_003F() & (1 << 5)) && (elIndustry.var_6_003F() & (1 << 4)))
+        {
+            unkE0C3B0 = indObj->var_28[elIndustry.var_6_003F() & 0x3];
+        }
+
+        uint32_t unkE0C3A4 = elIndustry.var_6_07C0();
+        uint32_t esi = indObj->var_3C[unkE0C3A4];
+        const auto baseHeight = elIndustry.baseHeight();
+        int16_t bbZOffset = baseHeight;
+
+        session.resetLastPS(); // Odd...
+        if (indObj->flags & IndustryObjectFlags::hasShadows)
+        {
+            if (session.getRenderTarget()->zoomLevel <= 1)
+            {
+                const auto shadowImageOffset = unkE0C3A4 * 4 + indObj->var_0E + rotation;
+                const ImageId shadowImage = baseColour.withIndex(shadowImageOffset).withTranslucency(Colours::getShadow(elIndustry.var_6_F800()));
```

Comment:
vanilla didn't have a check here for construction ghosts. Investigate if the old behaviour is required.

Reply:
Can't actually find an industry object with shadows.

Reply:
If its still actually possible to make one with shadows, should we support it?

Reply:
We have already written the code for it. I'm sure someone must have used it. I didn't do a super intensive search.